caldav/match.go

```diff
@@ -0,0 +1,224 @@
+		match, err := matchPropFilter(propFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func matchPropFilter(filter PropFilter, comp *ical.Component) (bool, error) {
+	// TODO: this only matches first field, can there be multiple like for CardDAV?
+	field := comp.Props.Get(filter.Name)
+	if field == nil {
+		return filter.IsNotDefined, nil
+	}
+
+	var zeroDate time.Time
+	if filter.Start != zeroDate {
+		match, err := matchPropTimeRange(filter.Start, filter.End, field)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+		for _, paramFilter := range filter.ParamFilter {
+			if !matchParamFilter(paramFilter, field) {
+				return false, nil
+			}
+		}
+	} else if filter.TextMatch != nil {
+		if !matchTextMatch(*filter.TextMatch, field.Value) {
+			return false, nil
+		}
+		for _, paramFilter := range filter.ParamFilter {
+			if !matchParamFilter(paramFilter, field) {
+				return false, nil
+			}
+		}
+		return true, nil
+	}
+	// empty prop-filter, property exists
+	return true, nil
+}
+
+func matchCompTimeRange(start, end time.Time, comp *ical.Component) (bool, error) {
+	// TODO what about other types of components?
+	if comp.Name != ical.CompEvent {
+		return false, nil
+	}
+	event := ical.Event{comp}
+
+	eventStart, err := event.DateTimeStart(start.Location())
+	if err != nil {
+		return false, err
+	}
+	eventEnd, err := event.DateTimeEnd(end.Location())
+	if err != nil {
+		return false, err
+	}
+
+	// Event starts in time range
+	if eventStart.After(start) && eventStart.Before(end) {
+		return true, nil
+	}
+	// Event ends in time range
+	if eventEnd.After(start) && eventEnd.Before(end) {
+		return true, nil
+	}
+	// Event covers entire time range plus some
+	if eventStart.Before(start) && eventEnd.After(end) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func matchPropTimeRange(start, end time.Time, field *ical.Prop) (bool, error) {
+	// The RFC says: "The CALDAV:prop-filter XML element contains a
+	// CALDAV:time-range XML element and the property value overlaps the
+	// specified time range".
+	// Not entirely sure how a property can "overlap" a time range, but I
+	// assume it means "the property is Date/Time that falls into the given
+	// time range.
+	ptime, err := field.DateTime(start.Location())
+	if err != nil {
+		return false, err
+	}
+	if ptime.After(start) && ptime.Before(end) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func matchParamFilter(filter ParamFilter, field *ical.Prop) bool {
+	// TODO there can be multiple values
+	value := field.Params.Get(filter.Name)
+	if value == "" {
+		return filter.IsNotDefined
+	} else if filter.IsNotDefined {
+		return false
+	}
+	if filter.TextMatch != nil {
+		return matchTextMatch(*filter.TextMatch, value)
+	}
+	return true
+}
+
+func matchTextMatch(txt TextMatch, value string) bool {
+	// TODO: handle text-match collation attribute
+	match := value == txt.Text
```

Comment:
According to https://datatracker.ietf.org/doc/html/rfc4791#section-9.7.5, text-match specifies as substring match.

Reply:
Oops, thanks, changed and added a test case for that.